olm/operator-registry.md

```diff
@@ -0,0 +1,291 @@
+    ex: `operator-registry delete-latest foo example.db`
+    
+*Reference non latest versions of bundles by image digests*
+
+Currently the operator-registry database has a table to the effect of:
+
+OperatorBundle
+    -name: foo
+    -bundle: "{$jsonblob}"
+    
+We will add a field to this table that will reference the bundle image path and, when the operator is not the latest version of the default channel, will not include the bundle blob itself:
+
+OperatorBundle
+    -name: foo
+    -bundle: NULL
+    -bundlePath: "quay.io/community-operators/foo:sha256@abcd123"
+    
+Additionally, we will update the grpc API that OLM currently uses to get the bundle that if it attempts to pull the non latest version will return null. In that case, it will make a second query to get the bundlePath. Then we will create a pod with that image that will serve the bundle data by writing the data to a configmap that can be read by OLM.
+
+In order to serve that manifest image, we cannot pull the image directly. Instead we will:
+
+1. Create a pod spec that uses an init container to inject a binary into the manifest image.
+2. Run the manifest image with that binary that writes to a configmap in a well known location
+3. Read the configmap data
+4. On return kill the pod
+5. Delete the configmap
+
+pod spec:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+  - name: operator-manifest
+    image: quay.io/community-operators/foo@sha256:abc123
+    command: ['operator-registry', 'serve' '-manifestdir "./manifests/"']
+  initContainers:
+  - name: init-manifest
+    image: quay.io/operator-framework/init-operator-manifest:latest
+    command: ['cp', 'operator-registry', '.']
+```
+
+We will also need to update operator-registry to include a new `serve` command that knows how to parse the manifest/ folder in the operator bundle image. This command will write to a configmap that OLM will use to get the bundle data.
+
+*DB Versioning*
+
+Since we are now calling this database schema a first class releasable object, we also need to be able to make changes to the schema over time. Along with that, we will need to be able to migrate to newer schema versions over time. To get started with that, we will update the existing operator-registry database to include a table that defines the current version of the database schema.
+
+*Make CatalogSource type for this image*
+
+Finally, once these operator registry images are servable, we will update the catalog source to be able to download these images and serve them.
+
+Additionally, we will update the spec to create a polling field that allows the index to be updated automatically on a configurable poll timer.
+
+```
+apiVersion: operators.coreos.com/v1alpha1
+kind: CatalogSource
+metadata:
+  name: example-catalog
+  namespace: olm
+spec:
+  sourceType: grpc
+  image: quay.io/operator-framework/community-operators:
+  pollTimer: 1m
```

Comment:
Perhaps this should be a block allowing additional config in the future;

```yaml
poll:
  interval: 1m
  startingAt: 09-20-2019T20:20:15
```

Reply:
This is a good point. Let's nest for now but leave the single field for now, which will make extending it easier in the future.